modules/calib3d/src/solvepnp.cpp

```diff
@@ -456,4 +456,249 @@ int solveP3P( InputArray _opoints, InputArray _ipoints,
+        double xi = objectPointsInCam.at<double>(0,0) / Zi;
+        double yi = objectPointsInCam.at<double>(1,0) / Zi;
+
+        s.at<double>(2*i,0) = xi;
+        s.at<double>(2*i+1,0) = yi;
+
+        L.at<double>(2*i,0) = -1 / Zi;
+        L.at<double>(2*i,1) = 0;
+        L.at<double>(2*i,2) = xi / Zi;
+        L.at<double>(2*i,3) = xi*yi;
+        L.at<double>(2*i,4) = -(1 + xi*xi);
+        L.at<double>(2*i,5) = yi;
+
+        L.at<double>(2*i+1,0) = 0;
+        L.at<double>(2*i+1,1) = -1 / Zi;
+        L.at<double>(2*i+1,2) = yi / Zi;
+        L.at<double>(2*i+1,3) = 1 + yi*yi;
+        L.at<double>(2*i+1,4) = -xi*yi;
+        L.at<double>(2*i+1,5) = -xi;
+    }
+}
+
+/**
+ * @brief The exponential map from se(3) to SE(3).
+ * @param twist A twist (v, w) represents the velocity of a rigid body as an angular velocity
+ * around an axis and a linear velocity along this axis.
+ * @param R1 Resultant rotation matrix from the twist.
+ * @param t1 Resultant translation vector from the twist.
+ */
+static void exponentialMapToSE3Inv(const Mat& twist, Mat& R1, Mat& t1)
+{
+    //see Exponential Map in http://ethaneade.com/lie.pdf
+    double vx = twist.at<double>(0,0);
+    double vy = twist.at<double>(1,0);
+    double vz = twist.at<double>(2,0);
+    double wx = twist.at<double>(3,0);
+    double wy = twist.at<double>(4,0);
+    double wz = twist.at<double>(5,0);
+
+    Mat rvec = (Mat_<double>(3,1) << wx, wy, wz);
+    Mat R;
+    Rodrigues(rvec, R);
+
+    double theta = sqrt(wx*wx + wy*wy + wz*wz);
+    double sinc = abs(theta) < 1e-8 ? 1 : sin(theta) / theta;
+    double mcosc = (abs(theta) < 1e-8) ? 0.5 : (1-cos(theta)) / (theta*theta);
+    double msinc = (abs(theta) < 1e-8) ? (1/6.0) : (1-sinc) / (theta*theta);
+
+    Matx31d dt;
+    dt(0) = vx*(sinc + wx*wx*msinc) + vy*(wx*wy*msinc - wz*mcosc) + vz*(wx*wz*msinc + wy*mcosc);
+    dt(1) = vx*(wx*wy*msinc + wz*mcosc) + vy*(sinc + wy*wy*msinc) + vz*(wy*wz*msinc - wx*mcosc);
+    dt(2) = vx*(wx*wz*msinc - wy*mcosc) + vy*(wy*wz*msinc + wx*mcosc) + vz*(sinc + wz*wz*msinc);
+
+    R1 = R.t();
+    t1 = -R1 * dt;
+}
+
+enum SolvePnPRefineMethod {
+    SOLVEPNP_REFINE_LM   = 0,
+    SOLVEPNP_REFINE_VVS  = 1
+};
+
+static void solvePnPRefine(InputArray _objectPoints, InputArray _imagePoints,
+                           InputArray _cameraMatrix, InputArray _distCoeffs,
+                           InputOutputArray _rvec, InputOutputArray _tvec,
+                           SolvePnPRefineMethod _flags,
+                           TermCriteria _criteria=TermCriteria(TermCriteria::EPS+TermCriteria::COUNT, 20, FLT_EPSILON),
+                           double _vvslambda=1)
+{
+    CV_INSTRUMENT_REGION();
+
+    Mat opoints_ = _objectPoints.getMat(), ipoints_ = _imagePoints.getMat();
+    Mat opoints, ipoints;
+    opoints_.convertTo(opoints, CV_64F);
+    ipoints_.convertTo(ipoints, CV_64F);
+    int npoints = std::max(opoints.checkVector(3, CV_32F), opoints.checkVector(3, CV_64F));
```

Comment:
After `opoints_.convertTo(opoints, CV_64F);` this check should always fail: `opoints.checkVector(3, CV_32F)`